pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,873 @@
+	if instanceInfoClone.Status == nil || len(instanceInfoClone.Status.ShareNames) == 0 {
+		klog.Warningf("trying to remove share %q from instanceInfo %q but it does not have Status subresource or no assigned shares", shareName, instanceInfoClone.Name)
+		return instanceInfoClone, false
+	}
+	newShareNames := make([]string, 0, len(instanceInfoClone.Status.ShareNames))
+	updated := false
+	for _, name := range instanceInfoClone.Status.ShareNames {
+		if name == shareName {
+			klog.V(6).Infof("found share %q in instanceInfo %q, exclude it in updated object", shareName, instanceInfoClone.Name)
+			updated = true
+			continue
+		}
+		newShareNames = append(newShareNames, name)
+	}
+	if updated {
+		instanceInfoClone.Status.ShareNames = newShareNames
+	}
+	return instanceInfoClone, updated
+}
+
+func (recon *multishareReconciler) instanceInfoNewCapacity(instanceInfoClone *v1alpha1.InstanceInfo) (*v1alpha1.InstanceInfo, bool) {
+	if instanceInfoClone.Status == nil || len(instanceInfoClone.Status.ShareNames) == 0 {
+		return instanceInfoClone, false
+	}
+	// if we don't know what's the step size, use the min instance size as step size.
+	stepSizeGb := util.MinMultishareInstanceSizeBytes
+	if instanceInfoClone.Status.CapacityStepSizeGb != 0 {
+		stepSizeGb = instanceInfoClone.Status.CapacityStepSizeGb
+	}
+	var targetInstanceSizeByte int64 = 0
+	for _, shareName := range instanceInfoClone.Status.ShareNames {
+		shareInfo, err := recon.shareLister.Get(shareName)
+		if err != nil {
+			klog.Warningf("error getting ShareInfo %q: %v", shareName, err)
+			continue
+		}
+		targetInstanceSizeByte += shareInfo.Spec.CapacityBytes
+	}
+	targetInstanceSizeByte = util.AlignBytes(targetInstanceSizeByte, util.GbToBytes(stepSizeGb))
+
+	// bound InstanceSizeByte to max and min of Multishare instance size
+	targetInstanceSizeByte = util.Max(targetInstanceSizeByte, util.MinMultishareInstanceSizeBytes)
+	targetInstanceSizeByte = util.Min(targetInstanceSizeByte, util.MaxMultishareInstanceSizeBytes)
+
+	if targetInstanceSizeByte == instanceInfoClone.Spec.CapacityBytes {
+		return instanceInfoClone, false
+	}
+	klog.Infof("updating instanceInfo %q capacity from %d to %d bytes in updated object", instanceInfoClone.Name, instanceInfoClone.Spec.CapacityBytes, targetInstanceSizeByte)
+	instanceInfoClone.Spec.CapacityBytes = targetInstanceSizeByte
+	return instanceInfoClone, true
+}
+
+func (recon *multishareReconciler) maybeRemoveInstanceInfoFinalizer(instanceInfo *v1alpha1.InstanceInfo) (*v1alpha1.InstanceInfo, error) {
+	if instanceInfo.DeletionTimestamp == nil {
+		klog.V(6).Infof("InstanceInfo %q doesn't have deletion timestamp, it shouldn't be deleted", instanceInfo.Name)
+		return instanceInfo, nil
+	}
+	if instanceInfo.Status == nil {
+		klog.Errorf("InstanceInfo %q marked to be deleted but Status is nil", instanceInfo.Name)
```

Comment:
return error?

Reply:
I agree that keeping the finalizer is safer. My reasoning is we should not get into this state where the deletetimestamp != nil and status == nil. (because in instanceEmpty() we return false for nil status).  It could be a user invoked delete. Is this reasoning valid?

klog.Warningf() may be a better log level?

Reply:
do you mean the user could manually mark the object for deletion?

Reply:
I think perhaps returning `instanceInfo` as is and `nil` err is better: 1) it keeps instanceInfo as is on the record for reconciler and 2) we already logged the Warning message, no need to log out some error message again